<a href="https://colab.research.google.com/github/BOrtizL/Proyecto-7-DataScience/blob/main/4_2_1_UDD_Proyecto_M7_Modelo_de_RandomForest_bc_BarbaraOrtiz.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

En este notebook aplicarmos el modelo de RandomForest en la data de baja confianza para tener una coparación de su rendimiento.

In [1]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [2]:
import pickle
import pandas as pd
import numpy as np

from sklearn.model_selection import train_test_split, cross_val_score
from sklearn.metrics import confusion_matrix, accuracy_score
from sklearn.preprocessing import StandardScaler
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import accuracy_score, classification_report
from sklearn.decomposition import PCA
from statsmodels.stats.outliers_influence import variance_inflation_factor
import matplotlib.pyplot as plt

**Cargar los datos de baja confianza transformados**

In [3]:
# Ruta del archivo original en Google Drive
ruta_archivo_drive = "/content/drive/My Drive/Proyecto 7/raw/df_concatenado_baja_confianza.pickle"

# Cargar el DataFrame desde el archivo
with open(ruta_archivo_drive, 'rb') as f:
    df_concatenado_baja_confianza = pickle.load(f)

In [4]:
# Paso 1: Separar las características y la etiqueta
X = df_concatenado_baja_confianza.drop('gender', axis=1)
y = df_concatenado_baja_confianza['gender']

In [5]:
# Paso 2: Dividir los datos en conjuntos de entrenamiento y prueba
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

In [6]:
# Paso 3: Preprocesamiento - Escalado de características
scaler = StandardScaler()
X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled = scaler.transform(X_test)


# Paso 4: Aplicar PCA en el conjunto de entrenamiento
pca = PCA(n_components=7)  # Número óptimo de componentes
X_train_pca = pca.fit_transform(X_train_scaled)
X_test_pca = pca.transform(X_test_scaled)

In [7]:
# Paso 5: Crear y entrenar el modelo RandomForest
rf_model = RandomForestClassifier(random_state=42)
rf_model.fit(X_train_pca, y_train)

# Paso 6: Predecir en el conjunto de prueba
y_pred = rf_model.predict(X_test_pca)

# Paso 7: Evaluar el modelo
print(classification_report(y_test, y_pred))

# Calcular la matriz de confusión
conf_matrix = confusion_matrix(y_test, y_pred)
print(conf_matrix)

              precision    recall  f1-score   support

           1       0.95      0.95      0.95        57
           2       0.88      0.88      0.88        26

    accuracy                           0.93        83
   macro avg       0.92      0.92      0.92        83
weighted avg       0.93      0.93      0.93        83

[[54  3]
 [ 3 23]]


**Observación:**

El modelo tiene un 93% de precisión, lo que significa que tienen un buen desempeño.

A continuación la interpretación de las métricas de clasificación:

- Precision:
Para la Clase 1, la precisión es del 95%, lo que significa que el 95% de las predicciones para la Clase 1 fueron correctas.
Para la Clase 2, la precisión es del 88%, lo que indica que el 88% de las predicciones para la Clase 2 fueron correctas.

- Recall:
Para la Clase 1, el recall es del 95%, lo que significa que el 95% de las muestras de la Clase 1 fueron identificadas correctamente.
Para la Clase 2, el recall es del 88%, lo que indica que el 88% de las muestras de la Clase 2 fueron identificadas correctamente.

- F1-score:
Para la Clase 1, el F1-score es del 95%.
Para la Clase 2, el F1-score es del 88%.

- Accuracy: La precisión global del modelo es del 93%, lo que significa que el 93% de las muestras fueron clasificadas correctamente.

- Matriz de Confusión:
El modelo clasificó correctamente 54 muestras de la Clase 1 y 23 muestras de la Clase 2.
Clasificó erróneamente 3 muestras de la Clase 1 y 3 muestras de la Clase 2.

En resumen, el modelo parece tener un buen rendimiento, con altas precisiones, recalls y un buen F1-score para ambas clases, así como una alta precisión global del 93%.

# **Conclusión:**

Ahora que tenemos el resultado de desempeño de ambos modelos, es el modelo de Regresión Logistica quien tiene un mejor performance (accuracy de 95% v/s 93%) y es el modelo que se sugiere avanzar en este proyecto.